### PR TITLE
[JUJU-4012] Fix apiserver watchers

### PR DESCRIPTION
**Pull request by SimonRichardson**
The changes to the watcher registry only ensured that the watchers
were workers. Once bootstrapped this caused all sorts of chaos. This
is because no watcher actually aligned with the new type of watcher from
the core package.

The solution was to grab the new FirstResult internal package and shove
it one level higher. Then we can use that from the apiserver package.
From there we then needed to remove all the references from the state
watcher package and use the core ones. Although they're kind of similar
they are in fact not quite the same in terms of behavior. We will have
to fix that once we move everything across.
For now, though, we should just force everything to be a core watcher and
be done with it.

There are virtually zero tests for the watcher package in apiserver and
considering this is the main driver of juju that is actually very
disappointing. We'll have to rectify that in the future.

This will unblock https://github.com/juju/juju/pull/15817 from landing.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Ensure that there are no errors or time outs from the uniter.

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
$ juju deploy ubuntu
$ juju debug-log -m controller
```

### CODE REVIEW

File: apiserver/common/relationunitswatcher.go [hunk 1]
```diff
@@ -19,9 +19,11 @@ type RelationUnitsWatcher interface {
 	watcher.Watcher[params.RelationUnitsChange]
 
 	// Stop is needed to implement facade.Resource.
+	// Deprecated: use Kill and Wait instead.
 	Stop() error
 
 	// Err implements watcher.Errer.
+	// Deprecated: use Kill and Wait instead.
 	Err() error
```
Comment (SimonRichardson) [inline apiserver/common/relationunitswatcher.go:27]: Leave these for now, there are a lot of tests that depend on `watcher.EnsureErr()` and we need to undo that. Just not in this PR.

File: core/pinger/pinger.go [hunk 2]
```diff
@@ -73,13 +73,31 @@ func (pt *Pinger) loop() error {
 	}
 }
 
-// NoopPinger implements the pinger interface but just does nothing
-type NoopPinger struct{}
+// NoopPinger implements the pinger interface, does nothing but waits for heat
+// death of the universe.
+type NoopPinger struct {
```
Comment (SimonRichardson) [inline core/pinger/pinger.go:78]: Now that all watchers have to behave like workers, you can't just return a thing that does nothing. It actually has to do something. This is the consequence of the new registry managing the lifecycle of the worker.

**Review by manadart — APPROVED** [review]
This is good, but we _must_ make sure to:
- Remove the usages of `watcher.EnsureErr()` and the methods it relies on.
- Remove `ValueResource` as an anti-pattern.

**Pull request merged**